### ISSUE TITLE
Recognize Type Expressions in the Context of Generics

### ISSUE THREAD

**Issue by max-muoto** [microsoft/pyright#8313]
Currently, Pyright can recognize valid type-expressions, take the following examples:

### Valid:

```python
from typing import Any

def klass_generator(klass: type[Any]) -> type[Any]:
   class Wrapper(klass):
      ...

   return Wrapper
 ```

### Invalid

```python
from typing import Any

def klass_generator(klass: str) -> type[Any]:
   # Expected type expression but received "str"
   class Wrapper(klass):
      ...

   return Wrapper
 ```

However, this doesn't apply to generics:

```python
from typing import Any


def klass_decorator[T: type[Any]](klass: T) -> T:
    # Expected type expression...
    class Wrapped(klass):
        ...

    return Wrapped
```

See this Pyright Playground [example](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoCCKcAUGQCYCmwUA1gDYCGAzswPpUDGujMuA2gBUAXLESV%2BROAF1pACgYtmowQEooAWgB8UESSgGonJqygB1EIwQJK5BSearh%2Bw66o02bVEhie5zSnpgABooACpGEDRmULCw2gB3SOinF1d0qAQQVBg5ACIAIWpcSjzVNIzDZgBXGxA5VQA6T29fNjkIqJjw%2BKSu8sr0rJz8gmAYShAyslcQShhqkBRzS2tbIA).

Mostly creating this to capture the fact that this is missing, but understandable if we want to avoid implementing support until there's more real feedback around it.

**Comment by erictraut** [comment]
Yes, this should work, so I consider this a bug. The wording of the diagnostic message is also misleading because this isn't a "type expression".

**Comment by max-muoto** [comment]
> Yes, this should work, so I consider this a bug. The wording of the diagnostic message is also misleading because this isn't a "type expression".

Thanks for the quick fix!

**Comment by erictraut** [comment]
BTW, I generally recommend using `type[T]` rather than specifying an upper bound of `type` for `T`. This tends to work more reliably across type checkers and use cases. You can use this as a workaround while you're waiting for the bug fix.

```python
from typing import cast

def klass_decorator[T](klass: type[T]) -> type[T]:
    class Wrapped(klass):
        ...
    return cast(type[T], Wrapped)
```

**Comment by max-muoto** [comment]
> BTW, I generally recommend using `type[T]` rather than specifying an upper bound of `type` for `T`. This tends to work more reliably across type checkers and use cases. You can use this as a workaround while you're waiting for the bug fix.
> 
> ```python
> from typing import cast
> 
> def klass_decorator[T](klass: type[T]) -> type[T]:
>     class Wrapped(klass):
>         ...
>     return cast(type[T], Wrapped)
> ```

Good to know, thanks!

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.371.
